pkg/service/dbaas/service.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/vshn/exoscale-metrics-collector/pkg/clients/exoscale"
+	"github.com/vshn/exoscale-metrics-collector/pkg/database"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	egoscale "github.com/exoscale/egoscale/v2"
+	db "github.com/vshn/exoscale-metrics-collector/pkg/database"
+	"github.com/vshn/exoscale-metrics-collector/pkg/service"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	groupVersionResources = map[string]schema.GroupVersionResource{
+		"pg": {
+			Group:    "exoscale.crossplane.io",
+			Version:  "v1",
+			Resource: "postgresqls",
+		},
+	}
+)
+
+// Detail a helper structure for intermediate operations
+type Detail struct {
+	Organization, DBName, Namespace, Plan, Zone string
+}
+
+// Context is the context of the DBaaS service
+type Context struct {
+	context.Context
+	dbaasDetails     []Detail
+	exoscaleDBaasS   []*egoscale.DatabaseService
+	aggregatedDBaasS map[db.Key]db.Aggregated
+}
+
+// Service is the context of the DBaaS service
+type Service struct {
+	exoscaleClient *egoscale.Client
+	k8sClient      dynamic.Interface
+	database       *db.DBaaSDatabase
+}
+
+// NewDBaaSService creates a Service with the initial setup
+func NewDBaaSService(exoscaleClient *egoscale.Client, k8sClient dynamic.Interface, databaseURL string) Service {
+	return Service{
+		exoscaleClient: exoscaleClient,
+		k8sClient:      k8sClient,
+		database: &db.DBaaSDatabase{
+			Database: db.Database{
+				URL: databaseURL,
+			},
+		},
+	}
+}
+
+// Execute executes the main business logic for this application by gathering, matching and saving data to the database
+func (s *Service) Execute(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Running metrics collector by step")
+
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.NewStep("Fetch cluster managed DBaaS", s.fetchManagedDBaaS),
+		p.NewStep("Fetch exoscale DBaaS usage", s.fetchDBaaSUsage),
+		p.NewStep("Filter supported DBaaS", filterSupportedServiceUsage),
+		p.NewStep("Aggregate DBaaS services by namespace and plan", aggregateDBaaS),
+		p.WithNestedSteps("Save to billing database", hasAggregatedInstances,
+			p.NewStep("Get billing date", s.getBillingDate),
+			p.NewStep("Save to database", s.saveToDatabase),
+		),
+	)
+
+	return p.RunWithContext(&Context{Context: ctx})
+}
+
+// fetchManagedDBaaS fetches instances from kubernetes cluster
+func (s *Service) fetchManagedDBaaS(ctx *Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	var dbaasDetails []Detail
+	for _, groupVersionResource := range groupVersionResources {
+		managedResources, err := s.k8sClient.Resource(groupVersionResource).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot list managed resource %s from cluster: %w", groupVersionResource.Resource, err)
+		}
+
+		for _, managedResource := range managedResources.Items {
+			dbaasDetail := Detail{
+				DBName: managedResource.GetName(),
+			}
```

Comment:
What happens if there are two databases with the same name (msql "foo" and postgres "foo")? Are DBaaS names unique across database types on exoscale?

And even if they are this is a potential problem.  Let's say tenant A has a mysql resource called "foo" and tenant B has a (not ready) postgres resource. We'll potentially bill tenant B for the mysql server right?

I know this is technically not an issue right now but If you start to implement this in a generic way it should actually also work when we handle multiple DBaaS

Reply:
In the first loop we use `GroupVersionResource` to request k8s resources by type, so in every iteration we only handle one type of database (postgres, mysql...)
I don't think the name is unique in exoscale, but when we save the information in the db, we don't even use it, we just take into account namespace, organization, database type and plan. So we can basically delete the name completely from here.
Thoughts?

Reply:
my 2 cents about the name uniqueness:

The names for the DBaaS instances are the same names as the managed resource on the cluster has. That name is unique for the whole cluster as the MR is cluster scoped and has a random appendix. Given that each cluster should have its own exoscale project, there shouldn't be any collisions.

Even if the exoscale project is shared accross different clusters, the chances for a collision are very minimal due to the random naming.

Reply:
But that doesn't solve the issue I presented right?  In `aggregateDBaaS` we match usage and details based on the name. Am I completely missing something?

Reply:
> Even if the exoscale project is shared accross different clusters, the chances for a collision are very minimal due to the random naming.

IMO we shouldn't rely on this being unlikely for actually billing resources..

Reply:
> But that doesn't solve the issue I presented right? In aggregateDBaaS we match usage and details based on the name. Am I completely missing something?

True, I haven't checked the code in detail yet, so I'm missing some context what exactly is wrong here.

Reply:
maybe we could use the k8s resource UID to differentiate between them. But, then we might need to add it in the facts somehow 🤔

Reply:
Don't we just need to match exoscale resources and managed resources not just on name but also on DB type?  I think that should completely solve the issue and would only need a very minimal change

Reply:
First of all the names of DBaaS are unique across organisations in Exoscale. It means you cannot have a postgres and a mysql instance called "db". You can test it fairly easy in Exoscale DBaaS tab by creating two different types of DBaaS. This fact alone insures that the type is irrelevant for our "key" and that the name is sufficient. Now we have an edge case, as @glrf pointed out, we would count 2 instances twice when the cluster has 2 resources with the same name but different DBaaS type (first in Ready state and second in error naturally because of the name constraint). For this edge case I update the code and now I'm matching not just the name but the the DBaaS type as well.